dev-docs/RFCs/vNext/layer-extension-rfc.md

```diff
@@ -0,0 +1,151 @@
+
+Secondly, because of the ad-hoc nature of layer customization, it is difficult to package and reuse an "extension" for other layers. This is apprent in the [kepler.gl repo](https://github.com/uber/kepler.gl/tree/master/src/deckgl-layers) where a custom version of every core layer must be created in order to override a simple default bahavior.
+
+To mitigate this, we can of course add as many customizable functionalities to the base layer as possible. However, it is a major concern that inflating the base layer will sacrefice performance in return. Additional attributes takes CPU cycles to generate and memory to store. Larger shaders take longer to compile. WebGL1 has a restricting limit of 16 attributes per vertex shader, so adding attributes without the user asking for it will make the layer less extensible instead.
+
+
+## Case Studies
+
+### fp64
+
+* Shader module: `project64` (replaces `project32`)
+
+### Data Filter
+
+See [Data Filter RFC](/dev-docs/RFCs/v6.0/data-filter-rfc.md)
+
+Components:
+
+* Props: `getFilterValue`, `filterRange`
+* Shader module: `filter`
+* Attributes: `instanceFilterValue`
+* Shader injection:
+  - Vertex shader: `filter_setVisibility(instanceFilterValue);`
+  - Frament shader: `gl_FragColor = filter_filterColor(gl_FragColor);`
+
+### Brushing
+
+See [ArcBrushingLayer example](https://gist.github.com/Pessimistress/dc2becf3809c67dc443b4dbab1b9a46f#file-index-html-L128)
+
+A generic version of this functionality would need the following components:
+
+* Props: `enableBrush`, `brushRadius`
+* Shader module: `brushing`
+* Shader injection:
+  - Vertex shader: `brush_setVisibility(instancePositions);`
+  - Frament shader: `gl_FragColor = brush_filterColor(gl_FragColor);`
+* Subscribe to hover events and store the mouse position in layer state
+
+## Proposal
+
+### Layer Class Changes
+
+Some changes to the base Layer class are needed to make a generic, reusable extension system work:
+
+* A base layer `getShaders` method that each layer can utilize to augment their shaders with extensions.
+* (Implemented in v7.1) Official `bufferLayout` support in attribute management. The `bufferLayout` state is used in PathLayer and PolygonLayer to describe the number of instance for each data object. Currently, `AttributeManager`'s auto-update feature always assumes 1:1 mapping between an instance and a data object. Without supporting variable layout, an extension will have to implement layer-specific updaters for these layers.
+
+### New LayerExtension Class
+
+Add a `LayerExtension` interface. All layer extensions should extend this class. It contains the following methods:
+
+- `getShaders(shaders)` - called after a layer's own `getShaders`, a hook to inject additional modules/code into the shaders
```

Comment:
`getShaders()` is a method because it was intended to be overridden. That does not seem to be the case for extensions. It would seem clearer to me to just replace this method with `vs`, `fs`, `modules` field. 

This would have the benefit (depending on one's point of view) of making this part of extensions look more like shader modules rather than layers.

Reply:
In the `fp64` extension case (if we decide to go down that route), the extension will remove a shader module from the layer (`project32`) and replace it with another (`project64`). Unless we implement some sort of exclusion system, this will result in a compiler error.

Reply:
> the extension will remove a shader module from the layer (project32) and replace it with another (project64).

Yes, I see.

I will say that it feels like this depends on a "module replacement system" that is assumed (but not directly described) in this RFC (which I see as mainly describing a "module injection" system). This makes the "replacement" parts of the fp64 proposal a little hard to review.

We could add more detail on this, or punt the replacement part of it to a second RFC/PR. But then we'd need to decide what to do with `getShaders` at this state.